compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirNotImplementedOverrideChecker.kt

```diff
@@ -0,0 +1,287 @@
+                } else {
+                    reporter.reportOn(source, FirErrors.ABSTRACT_CLASS_MEMBER_NOT_IMPLEMENTED, declaration, notImplemented, context)
+                }
+                alreadyReportedAbstractNotImplemented = true
+            }
+        }
+    }
+
+    private fun FirCallableDeclaration<*>.isFromInterface(context: CheckerContext): Boolean =
+        (getContainingClass(context) as? FirRegularClass)?.isInterface == true
+
+    private fun collectCallableMembers(firClass: FirClass<*>, context: CheckerContext): Collection<FirCallableDeclaration<*>> =
+        firClass.declarations.filterIsInstance<FirCallableDeclaration<*>>() +
+                dataClassMembers(firClass, context) +
+                delegatedMembers(firClass, context)
+
+    private val FirSimpleFunction.matchesEqualsSignature: Boolean
+        get() = valueParameters.size == 1 && valueParameters[0].returnTypeRef.coneType.isNullableAny &&
+                returnTypeRef.coneType.isBoolean
+
+    private val FirSimpleFunction.matchesHashCodeSignature: Boolean
+        get() = valueParameters.isEmpty() &&
+                returnTypeRef.coneType.isInt
+
+    private val FirSimpleFunction.matchesToStringSignature: Boolean
+        get() = valueParameters.isEmpty() &&
+                returnTypeRef.coneType.isString
+
+    private val FirSimpleFunction.matchesDataClassSyntheticMemberSignatures: Boolean
+        get() = (this.name == EQUALS_NAME && matchesEqualsSignature) ||
+                (this.name == HASHCODE_NAME && matchesHashCodeSignature) ||
+                (this.name == TOSTRING_NAME && matchesToStringSignature)
+
+    private val SYNTHETIC_NAMES = listOf(EQUALS_NAME, HASHCODE_NAME, TOSTRING_NAME)
+
+    // See [DataClassMembersGenerator#generate]
+    // But, this one doesn't create IR counterparts. We just need what synthetic members would be generated for data/inline classes.
+    private fun dataClassMembers(firClass: FirClass<*>, context: CheckerContext): Collection<FirCallableDeclaration<*>> {
+        if (firClass !is FirRegularClass || (!firClass.isData && !firClass.isInline)) return emptyList()
+
+        val contributedInThisType = firClass.declarations.mapNotNull {
+            if (it is FirSimpleFunction && it.matchesDataClassSyntheticMemberSignatures) {
+                it.name
+            } else
+                null
+        }
+
+        val contributedInSupertypes = mutableMapOf<Name, FirCallableDeclaration<*>>()
+        val classScope = firClass.unsubstitutedScope(context)
+        for (name in SYNTHETIC_NAMES) {
+            classScope.processFunctionsByName(name) {
+                val declaration = it.fir
+                if (declaration.matchesDataClassSyntheticMemberSignatures && declaration.modality != Modality.FINAL) {
+                    contributedInSupertypes.putIfAbsent(declaration.name, declaration)
```

Comment:
That's right that if the current `data` class has its own contributed, e.g., `toString`, it will be added to `contributedInSupertypes`, and then returned as synthetic data class members. Therefore, we won't create a fake override at the end. So, we can avoid reporting a false alarm on, e.g., `toString`.

Yeah, the naming seems not right, but it's originated from `DataClassMembersGenerator`, so I would not bother with that.